app/src/main/java/org/simple/clinic/scanid/ScanSimpleIdScreenController.kt

```diff
@@ -39,21 +43,40 @@ class ScanSimpleIdScreenController @Inject constructor(
         .map { scannedQrCode ->
           try {
             val bpPassportCode = UUID.fromString(scannedQrCode)
-            ScanSimpleIdScreenPassportCodeScanned.ValidPassportCode(bpPassportCode)
+            ValidPassportCode(bpPassportCode)
           } catch (e: IllegalArgumentException) {
-            ScanSimpleIdScreenPassportCodeScanned.InvalidPassportCode
+            InvalidPassportCode
           }
         }
 
     events.mergeWith(scannedBpPassportCodes)
   }
 
   private fun handleScannedBpPassportCodes(events: Observable<UiEvent>): Observable<UiChange> {
-    val scannedBpPassportCodeStream = events
-        .ofType<ScanSimpleIdScreenPassportCodeScanned.ValidPassportCode>()
+    val showKeyboardEvents = events
+        .ofType<ShowKeyboardEvent>()
+
+    val hideKeyboardEvents = events
+        .ofType<HideKeyboardEvent>()
+
+    val keyboardEvents = Observable
+        .merge(showKeyboardEvents, hideKeyboardEvents)
+
+    val sharedScannedBpPassportCodeStream = events
```

Comment:
Yes, that's because of lines `94` and `97` (in the new screen). We have to share the stream because the first subscription will dispose itself as soon as the keyboard event stream emits. Line `97` will ensure that we utilize QR code scans only when the keyboard is not shown on the screen. This stream relies on a lot of state (previous and current) across streams and I am unable to see how to provide a cue about this stateful event stream.

Reply:
Right. Let's revisit this after we complete the feature.